File: app/src/main/java/com/vaibhav/moviedetails/viewmodels/MovieListViewModel.java
package com.vaibhav.moviedetails.viewmodels;

import android.arch.lifecycle.ViewModel;
import android.content.Context;

import com.vaibhav.moviedetails.commons.LogUtils;
import com.vaibhav.moviedetails.contract.MovieListContract;
import com.vaibhav.moviedetails.data.Movie;
import com.vaibhav.moviedetails.data.MovieResponse;
import com.vaibhav.moviedetails.db.MoviesDatabase;
import com.vaibhav.moviedetails.network.ApiClient;
import com.vaibhav.moviedetails.network.ApiInterface;

import java.lang.ref.WeakReference;
import java.util.ArrayList;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * @author Vaibhav Bhandula on 09/10/18.
 */
public class MovieListViewModel extends ViewModel {

    private static final String TAG = MovieListViewModel.class.getSimpleName();

    private WeakReference<Context> weakReference;
    private MovieListContract movieListContract;

    private String query = "";
    private int page = 1;

    public MovieListViewModel(Context context, MovieListContract movieListContract) {
        weakReference = new WeakReference<>(context);
        this.movieListContract = movieListContract;
    }

    public void initViewModel() {
        setQuery("friends");
        getBookMarkedMovies();
    }

    public void setQuery(String query) {
        if (this.query.equalsIgnoreCase(query)) {
            return;
        }
        this.query = query;
        getMovies();
    }

    public void bookMarkMovie(Movie movie) {
        Observable.just(movie)
                .subscribeOn(Schedulers.io())
                .subscribe(new Observer<Movie>() {
                    @Override public void onSubscribe(Disposable d) {

                    }

                    @Override public void onNext(Movie movie) {
                        MoviesDatabase.getInstance(getContext()).getEventDao().insert(movie);
                    }

                    @Override public void onError(Throwable e) {
                        LogUtils.printLog(TAG, "error");
                    }

                    @Override public void onComplete() {
                        LogUtils.printLog(TAG, "added");
                    }
                });
    }

    public void removeBookMark(Movie movie) {
        Observable.just(movie)
                .subscribeOn(Schedulers.io())
                .subscribe(new Observer<Movie>() {
                    @Override public void onSubscribe(Disposable d) {

                    }

                    @Override public void onNext(Movie movie) {
                        MoviesDatabase.getInstance(getContext()).getEventDao().delete(movie);
                    }

                    @Override public void onError(Throwable e) {
                        LogUtils.printLog(TAG, "error");
                    }

                    @Override public void onComplete() {
                        LogUtils.printLog(TAG, "removed");
                    }
                });
    }

    private void getBookMarkedMovies() {
        Disposable disposable = MoviesDatabase.getInstance(getContext())
                .getEventDao()
                .getAllMovies()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(movies -> {
                    if (movieListContract != null) {
                        movieListContract.showBookMarkedMovies(new ArrayList<>(movies));
                    }
                });
    }

    private void getMovies() {
        ApiInterface apiInterface = ApiClient.getClient().create(ApiInterface.class);
        Observable<MovieResponse> call = apiInterface.getMovieList(page, query);
        call.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<MovieResponse>() {
                    @Override public void onSubscribe(Disposable d) {

                    }

                    @Override public void onNext(MovieResponse movieResponse) {
                        if (movieListContract != null && movieResponse != null) {
                            movieListContract.showSearchMovies(movieResponse.getMovies());
                        }
                    }

                    @Override public void onError(Throwable e) {
                        if (movieListContract != null) {
                            movieListContract.showError(e.getMessage());
                        }
                    }

                    @Override public void onComplete() {

                    }
                });
    }

    private Context getContext() {
        if (weakReference != null) {
            return weakReference.get();
        }
        return null;
    }
}

File: app/src/main/java/com/vaibhav/moviedetails/viewmodel/MovieListViewModel.kt
package com.vaibhav.moviedetails.viewmodel

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import android.os.Handler
import android.os.Looper
import com.vaibhav.moviedetails.data.Movie
import com.vaibhav.moviedetails.repository.MovieRepository
import io.reactivex.Observer
import io.reactivex.disposables.Disposable

/**
 * @author Vaibhav Bhandula on 08/04/19.
 */
class MovieListViewModel : ViewModel() {

    var moviesList = MutableLiveData<ArrayList<Movie>>()
    var bookMarkedList = MutableLiveData<ArrayList<Movie>>()
    private var moviesRepository = MovieRepository()

    private var query = ""
    private var pageNo = 1
    private var isEndOfList = false
    private var isQueryChanged = false

    init {
        setQuery("friends")
        getBookMarkedMovies()
    }

    fun setQuery(query: String) {
        if (this.query.equals(query, ignoreCase = true)) {
            return
        }
        this.query = query
        pageNo = 1
        isEndOfList = false
        isQueryChanged = true
        getMovies(false)
    }

    fun loadNextPage() {
        if (isEndOfList) {
            return
        }
        pageNo++
        getMovies(true)
    }

    fun addBookMark(movie: Movie) {
        moviesRepository.bookMarkMovie(movie)
    }

    fun deleteBookMark(movie: Movie) {
        moviesRepository.removeBookMark(movie, object : Observer<Boolean> {
            override fun onComplete() {
            }

            override fun onSubscribe(d: Disposable) {
            }

            override fun onNext(t: Boolean) {
                if (t) {
                    updateBookMarksInSearch()
                }
            }

            override fun onError(e: Throwable) {
            }

        })
    }

    private fun getMovies(pagination: Boolean) {
        moviesRepository.getMoviesList(pageNo, query, object : Observer<ArrayList<Movie>> {
            override fun onSubscribe(d: Disposable) {
            }

            override fun onComplete() {
            }

            override fun onNext(t: ArrayList<Movie>) {
                if (t.isEmpty()) {
                    isEndOfList = true
                } else {
                    checkForBookMarks(t, pagination)
                }
            }

            override fun onError(e: Throwable) {
                pageNo--
            }

        })
    }

    private fun getBookMarkedMovies() {
        moviesRepository.getBookMarkedMovies(object : Observer<ArrayList<Movie>> {
            override fun onSubscribe(d: Disposable) {
            }

            override fun onComplete() {
            }

            override fun onNext(t: ArrayList<Movie>) {
                bookMarkedList.value = t
                updateBookMarksInSearch()
            }

            override fun onError(e: Throwable) {
            }

        })
    }

    private fun checkForBookMarks(movies: ArrayList<Movie>, pagination: Boolean) {
        var oldList = moviesList.value ?: ArrayList()
        movies.forEach {
            val item = bookMarkedList.value?.find { bookmarked -> bookmarked.imdbId == it.imdbId }
            it.bookmarked = (item != null)
        }
        if (pagination) {
            oldList.addAll(movies)
        } else {
            oldList = ArrayList(movies)
        }
        moviesList.value = oldList
    }

    private fun updateBookMarksInSearch() {
        Handler(Looper.getMainLooper()).postDelayed({
            val movies = moviesList.value ?: ArrayList()
            movies.forEach {
                val item = bookMarkedList.value?.find { bookmarked -> bookmarked.imdbId == it.imdbId }
                if (item != null) {
                    it.bookmarked = item.bookmarked
                } else {
                    it.bookmarked = false
                }
            }
            moviesList.postValue(movies)
        }, 200)
    }
}
